components/raftstore/src/store/async_io/write.rs

```diff
@@ -0,0 +1,636 @@
+                for t in &task.request_times {
+                    metrics
+                        .before_write
+                        .observe(duration_to_sec(now.saturating_duration_since(*t)));
+                }
+            }
+        }
+    }
+
+    fn after_write_to_kv_db(&mut self, metrics: &StoreWriteMetrics) {
+        if metrics.waterfall_metrics {
+            let now = Instant::now();
+            for task in &self.tasks {
+                for t in &task.request_times {
+                    metrics
+                        .kvdb_end
+                        .observe(duration_to_sec(now.saturating_duration_since(*t)));
+                }
+            }
+        }
+    }
+
+    fn after_write_to_raft_db(&mut self, metrics: &StoreWriteMetrics) {
+        if metrics.waterfall_metrics {
+            let now = Instant::now();
+            for task in &self.tasks {
+                for t in &task.request_times {
+                    metrics
+                        .write_end
+                        .observe(duration_to_sec(now.saturating_duration_since(*t)))
+                }
+            }
+        }
+    }
+}
+
+struct Worker<EK, ER, T, N>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: Transport,
+    N: Notifier,
+{
+    store_id: u64,
+    tag: String,
+    engines: Engines<EK, ER>,
+    receiver: Receiver<WriteMsg<EK, ER>>,
+    notifier: N,
+    trans: T,
+    batch: WriteTaskBatch<EK, ER>,
+    cfg_tracker: Tracker<Config>,
+    raft_write_size_limit: usize,
+    metrics: StoreWriteMetrics,
+    message_metrics: RaftSendMessageMetrics,
+    perf_context: EK::PerfContext,
+}
+
+impl<EK, ER, T, N> Worker<EK, ER, T, N>
+where
+    EK: KvEngine,
+    ER: RaftEngine,
+    T: Transport,
+    N: Notifier,
+{
+    fn new(
+        store_id: u64,
+        tag: String,
+        engines: Engines<EK, ER>,
+        receiver: Receiver<WriteMsg<EK, ER>>,
+        notifier: N,
+        trans: T,
+        cfg: &Arc<VersionTrack<Config>>,
+    ) -> Self {
+        let batch = WriteTaskBatch::new(
+            engines.kv.write_batch_with_cap(KV_WB_DEFAULT_SIZE),
+            engines.raft.log_batch(RAFT_WB_DEFAULT_SIZE),
+        );
+        let perf_context = engines
+            .kv
+            .get_perf_context(cfg.value().perf_level, PerfContextKind::RaftstoreStore);
+        let cfg_tracker = cfg.clone().tracker(tag.clone());
+        Self {
+            store_id,
+            tag,
+            engines,
+            receiver,
+            notifier,
+            trans,
+            batch,
+            cfg_tracker,
+            raft_write_size_limit: cfg.value().raft_write_size_limit.0 as usize,
+            metrics: StoreWriteMetrics::new(cfg.value().waterfall_metrics),
+            message_metrics: Default::default(),
+            perf_context,
+        }
+    }
+
+    fn run(&mut self) {
+        let mut stopped = false;
+        while !stopped {
+            let mut handle_begin = None;
+
+            let mut first_time = true;
```

Comment:
A more clear way is calling `recv` and then loop `try_recv`.

Reply:
If so, the logic of handling msg will be written twice?

Reply:
Handling msg can be a method.

Reply:
Done